File: non_linear/binary_tree/binary_tree.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None
        self.parent = None

class BinaryTree:
    def __init__(self):
        self.root = None
        self.size = 0

    def buildTreeFromPreIn(self, preOrder, inOrder):
        if len(preOrder) == 0:
            return None

        root = Node(preOrder[0])
        rootIndex = inOrder.index(preOrder[0])

        root.left = self.buildTreeFromPreIn(preOrder[1:rootIndex+1], inOrder[:rootIndex])
        root.right = self.buildTreeFromPreIn(preOrder[rootIndex+1:], inOrder[rootIndex+1:])

        return root
    
    def buildTreeFromPostIn(self, post, ino):
        if len(post) == 0:
            return None

        root = Node(post[-1])
        rootIndex = ino.index(post[-1])

        root.left = self.buildTreeFromPostIn(post[:rootIndex], ino[:rootIndex])
        root.right = self.buildTreeFromPostIn(post[rootIndex:-1], ino[rootIndex+1:])

        return root
        
    
    def inOrder(self, node):
        if node is None:
            return []
        return self.inOrder(node.left) + [node.data] + self.inOrder(node.right)

    def preOrder(self, node):
        if node is None:
            return []
        return [node.data] + self.preOrder(node.left) + self.preOrder(node.right)
    
    def postOrder(self, node):
        if node is None:
            return []
        return self.postOrder(node.left) + self.postOrder(node.right) + [node.data]

preOrder = [1, 2, 4, 5, 3, 6, 7]
inOrder = [4, 2, 5, 1, 6, 3, 7]

treeA = BinaryTree()
treeA.root = treeA.buildTreeFromPreIn(preOrder, inOrder)

print("InOrder: ", treeA.inOrder(treeA.root))
print("PreOrder: ", treeA.preOrder(treeA.root))
print("PostOrder: ", treeA.postOrder(treeA.root))

print("====================================")

postOrder = [4, 5, 2, 6, 7, 3, 1]
inOrder = [4, 2, 5, 1, 6, 3, 7]

treeB = BinaryTree()
treeB.root = treeB.buildTreeFromPostIn(postOrder, inOrder)

print("InOrder: ", treeB.inOrder(treeB.root))
print("PreOrder: ", treeB.preOrder(treeB.root))
print("PostOrder: ", treeB.postOrder(treeB.root))

File: non_linear/binary_tree/BinaryTree.cpp
#include<iostream>
#include<queue>
#include<stack>

using namespace std;

#define MAX_SIZE 25000

typedef struct Node {
	int data;
	Node* left;
	Node* right;

	Node(int value) : data(value), left(nullptr), right(nullptr) {}
} Node;

class BinaryTree {

	Node* inserter(Node* root, int key) {
		Node* newnode = new Node(key);

		if (root == nullptr)
		{
			root = newnode;
			return root;
		}
		else {

			Node* q[MAX_SIZE];
			int f, r;
			f = r = -1;


			q[++r] = root;
			f = r = 0;

			while (f <= r) {
				Node* temp = q[f];
				f++;

				if (temp->left != NULL)
					q[++r] = temp->left;

				else {
					temp->left = newnode;
					return root;
				}

				if (temp->right != NULL)
					q[++r] = temp->right;
				else {
					temp->right = newnode;
					return root;
				}

			}


		}

	}

public:
	Node* root;
	BinaryTree() {
		root = nullptr;
	}

	void insert(int key) {

		root = inserter(root, key);

	}

	void inorderTraversal(Node* root) {
		if (root == NULL)
			return ;

		inorderTraversal(root->left);
		cout << root->data << " ";
		inorderTraversal(root->right);

	}

	void mirror(Node* node) {

		if (node->right)
			mirror(node->right);
		if (node->left)
			mirror(node->left);

		swap(node->left, node->right);

	}

	void leafProduct(Node* root, int &product) {
		if (root->left == NULL && root->right == NULL)
			product *= root->data;

		else {
			leafProduct(root->left, product);
			leafProduct(root->right, product);
		}
	}

	void leafSum(Node* root, int &summ) {
		if (root->left == NULL && root->right == NULL)
			summ += (root->data);

		else {
			leafSum(root->left, summ);
			leafSum(root->right, summ);
		}
	}
};

int main() {
	BinaryTree t;

	t.insert(1);
	t.insert(2);
	t.insert(3);


	int product = 1, summ = 0;
	t.leafProduct(t.root, product);
	cout << product;

	t.leafSum(t.root, summ);
	cout << summ;

	// t.inorderTraversal(t.root);

	return 0;
}